aiida_quantumespresso/parsers/pw2gw.py

```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import numpy as np
+from io import StringIO
+from aiida.common import exceptions
+from aiida.parsers import Parser
+
+from aiida_quantumespresso.calculations.pw2gw import Pw2gwCalculation
+
+class Pw2gwParser(Parser):
+    """`Parser` implementation for the `Pw2gwCalculation` calculation job class."""
+
+    def parse(self, **kwargs):
+        """Parse the retrieved files of a completed `Pw2gwCalculation` into output nodes.
+
+        Two nodes that are expected are the default 'retrieved' `FolderData` node which will store the retrieved files
+        permanently in the repository. The second required node is a filepath under the key `retrieved_temporary_files`
+        which should contain the temporary retrieved files.
+        """
+        try:
+            retrieved = self.retrieved
+        except exceptions.NotExistent:
+            return self.exit(self.exit_codes.ERROR_NO_RETRIEVED_FOLDER)
+        else:
+            retrieved_names = retrieved.list_object_names()
+
+        files = Pw2gwCalculation._internal_retrieve_list + [Pw2gwCalculation._DEFAULT_OUTPUT_FILE]
+        if any(not _ in retrieved_names for _ in files):
+            return self.exit(self.exit_codes.ERROR_OUTPUT_FILES)
+
+        for name in Pw2gwCalculation._internal_retrieve_list:
+            content = retrieved.get_object_content(name)
+
+            try:
+                data =  np.loadtxt(StringIO(content))
+            except ValueError:
+                return self.exit(self.exit_codes.ERROR_OUTPUT_FILES)
+            if not len(data.shape) == 2:
+                return self.exit(self.exit_codes.ERROR_OUTPUT_FILES)
+            if not data.shape[0] > 0 or not data.shape[1] == 2:
+                return self.exit(self.exit_codes.ERROR_OUTPUT_FILES)
```

Comment:
Looking a bit closer at the format of these output files and what you are doing in the parser, I am wondering two things:

1. If you are really only just validating the format of the files here, maybe we can make the exit code a bit more specific. Maybe `330 - ERROR_OUTPUT_FILES_INVALID_FORMAT - 'The eps*.dat output files do not have the expected shape (N, 2).' `
2. You are really just parsing these files to check that their content is valid. How are these files used in potential follow up calculations? Would it maybe make sense to turn one, or multiple in actual output nodes? Or do you think the additional cost on the database does not outweigh the benefit of having the data in node form, besides just as raw files in the retrieved folder.

Reply:
The most common post-processing to be applied to this files would be to apply a gaussian/lorentzian broadening or the KK-relations to get the real part of the dielectric function.

The size in .npy format of a possible output (using pw2gw default input values) file  would be around 25KB (I assume the size would be similar in the DB).
It could increase to close to 1MB depending on the energy range and sampling rate.

If you think it wouldn't be too cumbersome to store, I could add an output node 'eps' of type ArrayData containing the following:
{ 'energy': [...],    'epsX': [...],    'epsY': [...],    'epsZ': [...]}

Reply:
These are difficult but important decisions, but since I have never worked with `pw2gw` its difficult to judge. If you don't think it is useful to have the files as nodes, or at least the added weight to the database outweighs the benefits, it is better not to add them. Or if in doubt, we don't add them as an output node for now and if in the future it does seems like it is useful, we can always still add it then. That is easier then removing it again

Reply:
I decided to add the node output node `eps` as above described with the corresponding test.